keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -666,7 +773,1310 @@ ways.  One is that we are dispatching requests to be served rather
+accurate implementation would keep track of enough historical
+information to revise all that scheduling.
+
+The order of request dispatches in the real world is taken to be the
+order of request completions in the virtual world.  In the case of
+ties, round robin ordering is used starting from the queue that most
+closely follows the one last dispatched from.  Requests are dispatched
+as soon as allowed by that ordering, the concurrency bound in the
+problem statement, and of course not dispatching before arrival.
+
+
+#### Implementation of Fair Queuing for Server Requests with equal allocations and serial execution, technique and problems
+
+One of the key implementation ideas is taken from the original paper.
+Define a global meter of progress named `R`, and characterize each
+request's execution interval in terms of that meter, as follows.
+
+```
+R(t) = Integral[from tau=epoch to tau=t] mu_equal(tau) dtau
+S(i,j) = R(B(i,j))
+F(i,j) = R(E(i,j))
+```
+
+In the current implementation, that global progress meter is called
+"virtual time".
+
+The value of working with `R` values rather than time is that they do
+not vary with `mu` and `NOS`.  Compare the following two equations
+(the first is derived from above, the second from the first and the
+definitions of `R`, `S`, and `F`).
+
+```
+Integral[from tau=B(i,j) to tau=E(i,j)] mu_equal(tau) / width(i,j) dtau
+        = len(i,j)
+
+F(i,j) = S(i,j) + len(i,j) * width(i,j)
+```
+
+The next key idea is that when it comes time to dispatch the next
+request in the real world, (a) it must be drawn from among those that
+have arrived but not yet been dispatched in the real world and (b) it
+must be the next one of those according to the ordering in the
+real-world dispatch rule given in the problem statement --- recall
+that is increasing `E`, with ties broken by round-robin ordering among
+queues.  The implementation maintains a cursor into that order.  The
+cursor is a queue index (for resolving ties according to round-robin)
+plus a bifurcated representation of each queue.  A queue's
+representation consists of two parts: a set of requests that are
+executing in the real world, and a FIFO of requests that are waiting
+in the real world.  When it comes time to advance the cursor, the
+problem is to find --- among the queues with requests waiting in the
+real world --- the one whose oldest real-world-waiting request (the
+one at head of the FIFO) has the lowest `E`, with ties broken
+appropriately.  This is the only place where the `B` and `E` values
+have an effect on the real world, and this fact is used to prune the
+implementation as explained next.
+
+It is not necessary to explicitly represent the `B` and `E`, nor even
+the `S` and `F`, of _every_ request that exists in the implementation
+at a given moment.  For each queue with requests waiting in the real
+world, all that is really needed (to support finding the next request
+to dispatch) is the `F` of the oldest one of those requests.  The
```

Comment:
There is a small technical error here: `F` (now `r_finish_virtual`) has dimension `time * seats` rather than `time`.  Let us imagine it corrected, the essence of the problem remains.

The problem is that with gross over-estimates of a request's duration, and width being different for different requests, the product of duration-guess and width is no longer a constant --- and this can effectively give big penalties to queues with relatively wide requests at their head.

Note that this problem is most severe for the solutions in which a queue executes one request at a time in the virtual world.  In the solutions where a queue executes several requests concurrently in the virtual world, this problem happens to a lesser degree --- but it still happens.

A simple solution is to guess not the duration but rather the product of duration and seats.  A constant guess at work rather than duration removes the differential penalties due to guessing duration and non-uniform widths.

The re-raises the question of what the guess should be.  In the past I have expressed nervousness at guessing low.  Reconsidering that now, I think that should be OK.  My reasoning is that we already accept divergence from the ideal bounded by the requests currently executing.  Errors in the duration guesses are compensated at each completion and do not accumulate.

So the simplest solution would be to guess a duration of zero.

I have been glossing over the execution duration padding (AKA extra latency).  We could make the execution duration guess equal the duration padding.  We could make it zero.

Reply:
> So the simplest solution would be to guess a duration of zero.

Thinking of what would be the potential problem with underestimates: I think the problem is that we may initially choose from the queue we shouldn't choose in the ideal world. At the beginning it can be simply round-robin.
Which indeed doesn't sound that bad, because it should relatively quickly self-correct.

The only potential problem I see with using 0 is that we treat all requests independenly from their width equally. Which I think shouldn't be the case - if we have q1 with requests of width 1 and q2 withrequests with width 4, I would intuitively choose 4 more requests from q1.

As a result, i would prefer to use some value that is very low (to ensure that when multiplied by width will self-correct quickly (not after 60s as in the initial example) but non-zero.
If we would estimate len(i,j) initially as `1ms`, then even for requests with width=10, it will self-correct after 10ms or so, which sounds acceptable. 

Does that make sense?

Reply:
If the initial guess is for a constant amount of _work_ (rather than duration), and duration padding is not added, then it does not matter what the guess is --- using any constant yields the same behavior.

If the initial guess is a guess at _duration_ then a large guess can cause the problem that started this discussion.  I think a small value like you suggest would be OK.

Reply:
Making the initial guess a "duration guess" is more compatible with the code (which is already pretty complicated so if possible we should try avoid adding more). And if we would use a small value (like 1ms I mentioned above), I think the drawback of using it as duration as opposed to `amount of work` is relatively small, so I would personally go this way.